File: 289.game-of-life.js
/**
 * @param {number[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var gameOfLife = function(board) {
	var i = 0, j = 0;
	var nLive = 0;
	var nAliveCnt = 0;

	var matrixRow = board.length;
	var matrixColumn = board[0].length;

	var copyMatrix = new Array();

	for (i = 0; i < matrixRow; i++) 
	{
		copyMatrix[i] = new Array();

		for (j = 0; j < matrixColumn; j++)
		{
			if (board[i][j] == 1) 
			{
				nLive = nLive + 1;
			}
			copyMatrix[i][j] = board[i][j];
		}
	}

	for(i = 0; i < matrixRow; i++)
	{
		for(j = 0; j < matrixColumn; j++)
		{
			nAliveCnt = 0;

			if ((i-1 >= 0) && (j - 1 >=0) && (board[i-1][j-1] == 1))
			{
				nAliveCnt++;
			}

			if((i-1 >= 0) && (board[i-1][j] == 1))
			{
				nAliveCnt++;
			}

			if((i - 1 >= 0) && ( j + 1 < matrixColumn) && (board[i-1][j+1] == 1))
			{
				nAliveCnt++;
			}

			if(( j - 1 >= 0) && (board[i][j -1] == 1))
			{
				nAliveCnt++;
			}

			if(( j + 1 < matrixColumn) && (board[i][j+1] == 1))
			{
				nAliveCnt++;
			}

			if(( i + 1 < matrixRow) && ( j - 1 >=0) && (board[i+1][j-1] ==1))
			{
				nAliveCnt++;
			}

			if(( i + 1 < matrixRow) && (board[i+1][j] == 1))
			{
				nAliveCnt++;
			}

			if(( i + 1 < matrixRow) && ( j + 1 <matrixColumn) && (board[i+1][j+1] == 1))
			{
				nAliveCnt++;
			}

			if (board[i][j] == 1)
			{
				if ((nAliveCnt == 2) || (nAliveCnt == 3))
				{
					copyMatrix[i][j] = 1;
				}
				else
				{
					copyMatrix[i][j] = 0;
					nLive--;
				}
			}

			if(board[i][j] == 0)
			{
				if (nAliveCnt == 3)
				{
					copyMatrix[i][j] = 1;
					nLive++;
				}
			}
		}
	}

	for(i = 0; i < matrixRow; i++)
	{
		for (j =0; j <matrixColumn; j++)
		{
			board[i][j] = copyMatrix[i][j];
		}
	}

};

File: 289.game-of-life.py
#
# @lc app=leetcode id=289 lang=python3
#
# [289] Game of Life
#
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        nAliveCnt = 0
        boardRow = len(board)
        boardColumn = len(board[0])

        copyBoard = [[0 for col in range(boardColumn)] for row in range(boardRow)]
        
        for i in range(boardRow):
            for j in range(boardColumn):
                copyBoard[i][j] = board[i][j]

        for i in range(boardRow):
            for j in range(boardColumn):
                nAliveCnt = 0

                # Check top left neighbor
                if (i - 1 >= 0) and (j - 1 >= 0) and (board[i - 1][j - 1] == 1):
                    nAliveCnt += 1
                
                # Check top neighbor
                if (i - 1 >= 0) and (board[i - 1][j] == 1):
                    nAliveCnt += 1

                # Check top right neighbor
                if (i - 1 >= 0) and (j + 1 < boardColumn) and (board[i - 1][j + 1] == 1):
                    nAliveCnt += 1
                
                # Check left neighbor 
                if (j - 1 >= 0) and (board[i][j - 1] == 1):
                    nAliveCnt += 1
                
                # Check right neighbor 
                if (j + 1 < boardColumn) and (board[i][j + 1] == 1):
                    nAliveCnt += 1
                
                # Check bottom left neighbor 
                if (i + 1 < boardRow) and (j - 1 >= 0) and (board[i + 1][j - 1] == 1):
                    nAliveCnt += 1
                
                # Check bottom neighbor 
                if (i + 1 < boardRow) and (board[i + 1][j] == 1):
                    nAliveCnt += 1
                
                # Check bottom right neighbor 
                if (i + 1 < boardRow) and (j + 1 < boardColumn) and (board[i + 1][j + 1] == 1):
                    nAliveCnt += 1

                # Live
                if board[i][j] == 1:
                    if (nAliveCnt == 2 or nAliveCnt == 3):
                        copyBoard[i][j] = 1
                    else:
                        copyBoard[i][j] = 0
                    
                # Dead 
                if board[i][j] == 0:
                    if nAliveCnt == 3:
                        copyBoard[i][j] = 1

        # Recopy to board 
        for i in range(boardRow):
            for j in range(boardColumn):
                board[i][j] = copyBoard[i][j]

    
